### PR TITLE
Sync project file source files changes instantly

### LINKED ISSUE #3602: File re-ordering is slow
Re-ordering files in the Solution Explorer is unnecessarily slow - each movement causes the project to be reloaded with ComputeSourcesAndFlags. It's not necessary - we know what the new source file layout will look like so we should just use that instead.

Happy to take a look at this this week.

### PR DESCRIPTION
**Pull request by saul**
Fixes #3602 

Re-ordering/renaming/adding/removing files from a project is no longer subject to the "Computing sources and flags..." dialog - it's now instant:

file move

### CODE REVIEW

**Comment by cartermp** [comment]
The assertion I'm seeing could actually be #3596 - @KevinRansom @dsyme does that sound right?

At any rate, it is instant when I try ordering files @saul, so assuming that the assertion I saw was #3596 then this has a 👍 from me.

**Comment by vasily-kirichenko** [comment]
I saw this assertion yesterday when testin the parsing-out-of-reactor PR. I thought I passed wrong parsing options. And yes, it was on Paket solution.

**Comment by cartermp** [comment]
Interesting.  I only got the assertion on this PR.  If I'd go back to master, it wouldn't appear.  I'll try again later today.

**Comment by majocha** [comment]
@cartermp "Parsing options don't contain any file" was added in the parsing improvements PR, maybe you have it merged?

**Comment by cartermp** [comment]
I removed F# support from my machine prior to testing out the other PR.  However, I've got multiple builds of VS on my machine and could have something borked up.  I'll verify again after sweeping off the VSs on my machine, but assuming that nobody else sees this, it should be good.

File: vsintegration/src/FSharp.ProjectSystem.FSharp/Project.fs [hunk 1]
```diff
@@ -1323,6 +1323,26 @@ namespace rec Microsoft.VisualStudio.FSharp.ProjectSystem
             member x.GetCompileItems() = let sources,_ = sourcesAndFlags.Value in sources
             member x.GetCompileFlags() = let _,flags = sourcesAndFlags.Value in flags
 
+            /// Sync the source files with what's listed in the Solution Explorer.
+            /// This is called when a file is added, moved or removed from the project.
+            member this.SyncWithHierarchy() =
+                let sources =
+                    this.AllDescendants
+                    |> Seq.choose (function
+                        | :? FileNode as fileNode when fileNode.ItemNode.ItemName = ProjectFileConstants.Compile ->
```
Comment (dsyme) [inline vsintegration/src/FSharp.ProjectSystem.FSharp/Project.fs]: Does this include generated items produced by MSBuild tasks, e.g. pars.fs, lex.fs, assemblyinfo files, resources etc.?

Comment (saul) [reply vsintegration/src/FSharp.ProjectSystem.FSharp/Project.fs]: Can you give me an example project where I can test this?

Comment (KevinRansom) [reply vsintegration/src/FSharp.ProjectSystem.FSharp/Project.fs]: Microsoft.Compiler.Private.dll.

Or anything with generated attributes.

Comment (KevinRansom) [reply vsintegration/src/FSharp.ProjectSystem.FSharp/Project.fs]: Hmmm it's FSharp.Compiler.Private.dll of course, because we changed the name.

Comment (saul) [reply vsintegration/src/FSharp.ProjectSystem.FSharp/Project.fs]: Ah thank you - in which case no this doesn't work. I'll reimplement it to be a bit more clever 👍

**Comment by cartermp** [comment]
I see a different assertion now on a clean box/VS install when opening `PublicAPI.fs` in Paket.sln in a debug hive:

assert-saul-pr

File: vsintegration/src/FSharp.ProjectSystem.FSharp/Project.fs (hunk 1 above)
Comment (dsyme) [reply vsintegration/src/FSharp.ProjectSystem.FSharp/Project.fs]: Super, thanks @saul, I'm glad I wasn't imagining something :)

**Comment by saul** [comment]
@dsyme @cartermp I've reimplemented the logic to infer the new compile sources. If the length of the inferred compile sources and the old compile sources doesn't match up, we fallback to the old ComputeSourcesAndFlags behaviour.

Note that I made a lot of whitespace changes in Project.fs - to see the diff without the whitespace, check this:
https://github.com/Microsoft/visualfsharp/pull/3604/files?w=1

**Comment by abelbraaksma** [comment]
Can this be tested or merged? Looks like an excellent and much awaited addition ;)

**Comment by saul** [comment]
@abelbraaksma there's still a few TODO items in the code that I need to fix. I'm also about 80% of the way to being able to move multiple files/folders instead of having to do them one at a time 😄

**Comment by abelbraaksma** [comment]
That would be awesome to have too! Unfortunately my knowledge of VS extensions is not big enough to be of any help. But perhaps someone else can have a look at the select issue?

If you need me to test something, let me know.

**Comment by saul** [comment]
@dotnet-bot test this please

**Comment by forki** [comment]
@saul can you please rebase?

**Comment by cartermp** [comment]
Renaming as per this:

> there's still a few TODO items in the code that I need to fix. I'm also about 80% of the way to being able to move multiple files/folders instead of having to do them one at a time

**Comment by saul** [comment]
@dotnet-bot test Windows_NT Release_ci_part1 please

**Comment by saul** [comment]
@dotnet-bot test Ubuntu14.04 Release_fcs please

**Comment by dsyme** [comment]
@saul This looks worth keeping alive and things haven't diverged too much - could you integrate with master?  Also reduce the diff since there are a bunch of unrelated changes?

**Comment by KevinRansom** [comment]
These look like good changes to have in the product.

Please reactivate when you get some time to devote to this.

Thank you

Kevin

**Pull request closed without merging**